src/app/api/auth/lib/nextauth-options.ts

```diff
@@ -88,39 +150,78 @@ export const nextAuthOptions: AuthOptions = {
       return false
     },
     session: async ({ session, token }) => {
-      // TODO: Need to figure out how to do this more efficiently
-      // Check if the user has a valid accessToken
-      // const validSession = await verifySession(token?.accessToken as string);
+      if (token) {
+        session.user.name = token.name as string
+        session.user.image = token.image as string
+        session.user.email = token.email as string
+        session.user.accessToken = token.accessToken as string
+        session.expires = new Date(
+          token.accessTokenExpires as number,
+        ).toISOString()
+        session.error = token.error
+      }
 
-      // if (!validSession) {
-      //   return undefined as any;
-      // }
+      return session
+    },
+    jwt: async ({ token, user, account }) => {
+      // Initial sign in
+      if (user && account) {
+        authLogger.debug('Initial sign in')
+        console.log('account', account)
+        token = {
+          accessToken: account.access_token as string,
+          // Access token expiration is provided as number in seconds since epoch (value is always 8 hours)
+          accessTokenExpires: (account.expires_at as number) * 1000,
+          refreshToken: account.refresh_token as string,
+          // Refresh token expiration is provided as number of seconds until expiration (value is always 6 months)
+          refreshTokenExpires:
+            Date.now() + (account.refresh_token_expires_in as number) * 1000,
+          ...user,
+        }
+
+        return token
+      }
 
-      // This is fine when the session is invalid
-      if (!token) {
-        return undefined as any
+      // Return previous token if the access token has not expired yet
+      if (
+        token.accessTokenExpires &&
+        Date.now() < (token.accessTokenExpires as number)
+      ) {
+        authLogger.debug('Access token valid')
+        return token
       }
 
-      // creates a new date that is 12 hours from now
-      const twelveHoursFromNow = new Date(
-        new Date().getTime() + 12 * 60 * 60 * 1000,
-      )
+      authLogger.debug('Access token has expired')
 
-      if (session.expires && new Date(session.expires) > twelveHoursFromNow) {
-        session.expires = twelveHoursFromNow.toISOString()
+      // Return previous token if the refresh token has expired
+      if (
+        token.refreshTokenExpires &&
+        Date.now() > (token.refreshTokenExpires as number)
```

Comment:
Is there a hole in this logic where `Date.now()` is exactly equal? Unlikely maybe since measured in seconds but could we make the refresh  case `>=`
```suggestion
        Date.now() >= (token.refreshTokenExpires as number)
```